### PR TITLE
Future warnings for hookspec changes AND support defaults to enable deprecation

### PR DESCRIPTION
**Pull request by tgoodlet**
As per discussion with @RonnyPfannschmidt I'm putting this draft up to start looking at dealing with #15. 

I've broke some of the tests due to a change to `varnames()` which now returns both the `arg` and `kwarg` names for a function.

There's also some other changes that I didn't fully revert that pertained to trying to validate `kwargs` that we might decide to declare in `hookspecs` as a way to define defaults when a hook call doesn't provide them.

### CODE REVIEW

File: pluggy.py [hunk 1]
```diff
@@ -659,7 +688,9 @@ def varnames(func):
         return ()
 
     args, defaults = spec.args, spec.defaults
-    args = args[:-len(defaults)] if defaults else args
+    index = -len(defaults) if defaults else None
+    args, defaults = args[slice(0, index)], args[slice(
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: too fancy of an xpression

File: pluggy.py [hunk 2]
```diff
@@ -477,14 +489,30 @@ def _verify_hook(self, hook, hookimpl):
...
+            raise PluginValidationError(
+                "Plugin %r for hook %r\nhookimpl definition: %s\n%s"
+                "Positional args {0} are declared in the hookimpl but "
+                "can not be found in the hookspec" %
+                (hookimpl.plugin_name, hook.name,
+                 _formatdef(hookimpl.function), notinspec)
+            )
+
+        # warn about new/upcoming arguments to the spec
+        if notinimpl:
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: that one is explicitly breaking what we use pluggy for (as in not naming all arguments if you dont need thme all)

Comment (tgoodlet) [inline pluggy.py]: I still find this weird.
Is it supposed to be a feature that you don't have to match the spec?

File: pluggy.py (hunk 1 above)
Comment (RonnyPfannschmidt) [inline pluggy.py]: the expression is magical enough that i have to think very hard to follow it by now
and this can be a very simple if .. else  thats really easy to follow

File: pluggy.py [hunk 3]
```diff
@@ -477,14 +489,30 @@ def _verify_hook(self, hook, hookimpl):
...
-                    _formatdef(hookimpl.function), ", ".join(hook.argnames)))
+        # positional arg checking
+        hian, han = set(hookimpl.argnames), set(hook.argnames)
+        notinspec = hian - han
+        notinimpl = han - hian
+
+        if notinspec:
+            raise PluginValidationError(
+                "Plugin %r for hook %r\nhookimpl definition: %s\n%s"
+                "Positional args {0} are declared in the hookimpl but "
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: this one at least from seeing the code is more confusing than before (this may be a miss-impression)

File: pluggy.py [hunk 4]
```diff
@@ -591,24 +619,25 @@ class _MultiCall:
...
-        self.kwargs = kwargs
-        self.kwargs["__multicall__"] = self
-        self.specopts = specopts
+        self.caller_kwargs = kwargs  # come from _HookCaller.__call__()
+        self.caller_kwargs["__multicall__"] = self
+        self.specopts = hook.spec_opts if hook else specopts
 
     def execute(self):
-        all_kwargs = self.kwargs
+        caller_kwargs = copy.copy(self.caller_kwargs)
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: at first glance this looks incomplete, was a .update call planned?

File: pluggy.py [hunk 5]
```diff
@@ -104,7 +114,7 @@ def setattr_hookspec_opts(func):
             if historic and firstresult:
                 raise ValueError("cannot have a historic firstresult hook")
             setattr(func, self.project_name + "_spec",
-                   dict(firstresult=firstresult, historic=historic))
+                    dict(firstresult=firstresult, historic=historic))
```
Comment (tgoodlet) [inline pluggy.py]: Yep of course.

File: pluggy.py (hunk 3 above)
Comment (tgoodlet) [inline pluggy.py]: Oh shoot I left in the format string syntax by accident...
The only thing I wanted to improve was reporting **all** the incompatible args in a single error instead of requiring that user runs multiple times to figure them all out.

File: pluggy.py (hunk 4 above)
Comment (tgoodlet) [inline pluggy.py]: @RonnyPfannschmidt yes this was from when I was trying to approach this by calling hooks with `func(**kwargs)` and popping undeclared defaults beforehand.
I'll probably remove it unless we end up finding another use for it.

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt bump. 
Tried to fix up most of the things you asked.
The part in `varnames()` where I parse out the `kwarg` names didn't change much (still have to use the `slice()`). Not sure how you were envisioning I simplify it further.

File: pluggy.py [hunk 6]
```diff
@@ -477,14 +489,18 @@ def _verify_hook(self, hook, hookimpl):
...
-        for arg in hookimpl.argnames:
-            if arg not in hook.argnames:
-                raise PluginValidationError(
-                    "Plugin %r\nhook %r\nargument %r not available\n"
-                    "plugin definition: %s\n"
-                    "available hookargs: %s" %
-                    (hookimpl.plugin_name, hook.name, arg,
-                    _formatdef(hookimpl.function), ", ".join(hook.argnames)))
+        # positional arg checking
+        hian, han = set(hookimpl.argnames), set(hook.argnames)
```
Comment (tgoodlet) [inline pluggy.py]: I'm thinking this should all be just one line in the if statement.

File: pluggy.py [hunk 7]
```diff
@@ -689,8 +677,8 @@ def varnames(func):
 
     args, defaults = spec.args, spec.defaults
     index = -len(defaults) if defaults else None
-    args, defaults = args[slice(0, index)], args[slice(
-        index, None if index else 0)]
+    args = args[:index]
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: that leads to always empty defaults due to replacing args and is till no if/else without a slice all

File: pluggy.py [hunk 8]
```diff
@@ -676,9 +675,11 @@ def varnames(func):
...
 
     args, defaults = spec.args, spec.defaults
-    index = -len(defaults) if defaults else None
-    args = args[:index]
-    defaults = args[slice(index, None if index else 0)]
+    if defaults:
+        index = -len(defaults)
+        args, defaults = args[:index], args[index:]
+    else:
+        defaults = []
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: well done
btw, shouldn't those be hash-able? (as in made into  tuple)

Comment (tgoodlet) [inline pluggy.py]: @RonnyPfannschmidt we do hash them per function already in `varnames()` no?

Comment (tgoodlet) [inline pluggy.py]: ohhh you mean each list of names. Yeah I can add that.

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt @hpk42 bump!

I think this adds what you originally requested @RonnyPfannschmidt but I personally think it's quite hideous, ill performs, and is confusing given normal function call semantics.

I still have to add tests of course.

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt @hpk42 rebased onto the new test reorg changes.

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt @hpk42 alright I added a test which shows how this all works now.

I still think the implementation of `_MultiCall.execute()` would be better implemented like I suggested quite a while ago using `func(**kwargs)`.

I'd like you guys to take a look before I do any more changes since I'm pretty sure this PR's current state matches the original requirements of #15.

**Comment by nicoddemus** [comment]
Just a gentle ping, what's the status of this PR?

**Comment by tgoodlet** [comment]
@nicoddemus well it's in dire need of thorough review :)

I have [my own criticisms of this particular approach](https://github.com/pytest-dev/pluggy/issues/15#issuecomment-268868136) but I wanted to at least get something going for #15.

I know @RonnyPfannschmidt [mentioned that](https://github.com/pytest-dev/pluggy/issues/15#issuecomment-268815910) he thought a test might be missing.

File: pluggy.py [hunk 9]
```diff
@@ -724,6 +745,8 @@ def remove(wrappers):
                 raise ValueError("plugin %r not found" % (plugin,))
 
     def _add_hookimpl(self, hookimpl):
+        """A an implementation to the callback chain.
```
Comment (tgoodlet) [inline pluggy.py]: s/A/Add

**Comment by tgoodlet** [comment]
Declining this as I have reintroduced the change set using two smaller (and more focused) PRs #42 and #43 to simplify the review process.
